# Snakemake workflow: `sm-SNIPER`

![Snakemake](https://img.shields.io/badge/snakemake-â‰¥6.3.0-brightgreen.svg)


A Snakemake workflow for **Highly accurate Single Nucleotide polymorphisms calling and Implication of haplotypes in Probe-capture based long-Read Nanopore sequencing**


## Software

## Dependencies 

#### Conda
#### Singularity >= 3.8.5

## Features

## Usage

The usage of this workflow is described in the [Snakemake Workflow Catalog](https://snakemake.github.io/snakemake-workflow-catalog/?usage=<owner>%2F<repo>).

## Configuration

### Execution

#### 1. Install and activate conda environment
It is recommended to execute always from within top level of the pipeline directory (i.e sm-SNIPER/)
#### 2. Execute a dry-run
Checking the pipeline with dry-run options. It is to print print a summary of the DAG of jobs
```
snakemake -p -n
```
#### 3. Execute workflow local
#### 4. Execute workflow on a cluster


## Results

## Tips